### PR TITLE
Acknowledge reception of data in `TrinoResult`

### LINKED ISSUE #232: `execute` does not kick off query on Trino Server
### Expected behavior

The call to `client.execute` does not start the query on the Trino cluster.

### Actual behavior

The query only starts on the Trino cluster once `client.fetch` is called.

### Steps To Reproduce

Not sure if this is an issue or just expected behavior but I noticed that the query is not actually run by the Trino cluster until data is requested by the client.

So this code will not actually run the query
```python
from trino.dbapi import connect

conn = connect(
    host="<host>",
    port=<port>,
    user="<username>",
    catalog="<catalog>",
    schema="<schema>",
)
cur = conn.cursor()
cur.execute("CREATE TABLE hive.<schema>.<name> AS SELECT 'a' AS a")
```
Nothing happens on the Trino server at this point.

The query only runs when we call fetch. This line is what kicks off the query.
```python
cur.fetch()
```

Basically it seems like the initial post request does not actual kick off the query. This seems unintuitive to me. Since I would have thought that execute would actually run the query. As an example if I am trying to insert data into a table, I don't feel the need to call `fetch` because there is no data to fetch. I just want the query to run and I want to be able to wait until it is successful. I guess I'm wondering what the reasoning is behind the current behavior. I believe the behavior of PyHive was to actually start the query when `execute` was called. Thanks!

### Log output

_No response_

### Operating System

macOS Monterey

### Trino Python client version

0.309.0

### Trino Server version

388

### Python version

 3.8.6

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by mdesmet**
## Description

Fixes #232, #95

Ensures the received data is properly acknowledged by calling the next_uri. This will avoid seeing failed queries in the query log when executing scalar queries as in the following example.

```
cur.execute("SELECT VERSION()")
cur.fetchone()
cur.cancel()
```

## Release notes

( ) This is not user-visible and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
## Breaking Changes

* Make the `execute` method of the cursor block until at-least one row is received.
  This means users no longer need to call `fetchone` or `fetchall` to make sure query
  actually starts executing on the Trino server. Note that results still need to be consumed
  by calling `fetchone` or `fetchall` to make sure query isn't considered idle and terminated
  on the server. ([#232](https://trinodb/trino-python-client/issues/232))
```

```markdown
* Properly propagate query failures to the client when using `fetchone`.
  ([#95](https://trinodb/trino-python-client/issues/95))
* Fix queries returning a single row from sometimes appearing as failed on the server.
  ([#220](https://trinodb/trino-python-client/issues/220))
```

### CODE REVIEW

File: trino/client.py [hunk 1]
```diff
@@ -603,20 +604,19 @@ def rownumber(self) -> int:
...
-            yield self._map_row(self._experimental_python_types, row, self._query.columns)
-        self._rows = None
-
-        # Subsequent fetches from GET requests until next_uri is empty.
-        while not self._query.finished:
-            rows = self._query.fetch()
+        # A query only transitions to a FINISHED state when the results are fully consumed:
+        # The reception of the data is acknowledged by calling the next_uri before exposing the data through dbapi.
+        while not self._query.finished or self._rows is not None:
+            rows = self._rows
```
Comment (nineinchnick) [inline trino/client.py:610]: `rows` is not needed, it would be enough to iterate over `_row`, especially if it's only assigned at the end of the loop.

**Review by nineinchnick — APPROVED** [review]

**Comment by mdesmet** [comment]
The CI issues can be simulated by following code. Seems that although the result has returned the Trino API still provides one or more next_uri's to fetch.

```
def test_query_cancellation_not_triggered(trino_connection):
    count_not_finished = 0
    for _ in range(0, 1000):
        cur = trino_connection.cursor()
        cur.execute("SELECT VERSION()")
        cur.fetchone()
        if not cur._query.finished:
            count_not_finished += 1

    print(str(count_not_finished) + " unfinished queries")
```

Some questions:

- can we handle this on the serverside to not let this type of queries fail if all results have been consumed?
- Why do we mark a cancelled query as failed if it's a valid use case to only retrieve a number of results and bail out?

**Comment by hashhar** [comment]
> Why do we mark a cancelled query as failed if it's a valid use case to only retrieve a number of results and bail out?

It's not valid. It's a hack that BI tools and clients use instead of limiting their query to pull only what is needed.
e.g. As a cluster admin it's very useful to see clients who run a query that returns billions of rows but just take 100 rows and leave the query hanging (instead of either cancelling or consuming results) which means that the output buffer on the server (and other processes) keeps occupying memory until the query times out.

> can we handle this on the serverside to not let this type of queries fail if all results have been consumed?

You know that `version()` would return one row but the server does not since results are streamed back to coordinator from workers and coordinator can't know that there isn't more data coming until it asks workers about it.

Change your experiment to a query which returns arbitrary number of rows and then you can't know anymore whether query is finished or not. Special casing the client protocol for queries which return single row doesn't seem useful.

**Comment by hashhar** [comment]
The alternative is to have a client protocol which is based on persistent TCP connections instead of HTTP long polling - which brings it's own set of problems.

**Comment by hashhar** [comment]
Also, specifically on why you cannot assume an empty rows being returned from API as proof that query has finished is the pipelined execution model. In queries it's possible for Trino to perform both table scans and output results at the same time. e.g. If you have a long chain of UNION ALL statements then Trino can start returning results as soon as the first UNION ALL query is done while other parts of the query are still executing .

This means that the client might observe periods of time where there is no data returned but a nextUri is still included. If the client were to assume no data == query finished then it'll drop any upcoming rows that will be produced.

**Comment by nineinchnick** [comment]
@hashhar would you agree that to address the original issue we should add a fetch call after `scalar()` to drain the cursor? We might also want to document this somewhere. I don't think we should try to make the driver too smart to work around these protocol limitations.

File: trino/client.py [hunk 2]
```diff
@@ -594,7 +594,8 @@ class TrinoResult(object):
 
     def __init__(self, query, rows=None, experimental_python_types: bool = False):
         self._query = query
-        self._rows = rows or []
+        # Initial rows from the first POST request
```
Comment (hashhar) [inline trino/client.py:597]: There's no guarantee initial rows will be present

File: trino/client.py [hunk 3]
```diff
@@ -603,20 +604,17 @@ def rownumber(self) -> int:
...
-            yield self._map_row(self._experimental_python_types, row, self._query.columns)
-        self._rows = None
-
-        # Subsequent fetches from GET requests until next_uri is empty.
-        while not self._query.finished:
-            rows = self._query.fetch()
-            for row in rows:
+        # A query only transitions to a FINISHED state when the results are fully consumed:
+        # The reception of the data is acknowledged by calling the next_uri before exposing the data through dbapi.
+        while not self._query.finished or self._rows is not None:
```
Comment (hashhar) [inline trino/client.py:609]: the or is wrong. rows can be missing while the query is not yet finished.

**Comment by hashhar** [comment]
@nineinchnick I don't agree 100%. It's still useful to make the client as smart as the JDBC driver where the implementation detail of the Trino REST protocol isn't visible to users.

But yes, since this might take more time than the quickfix I think we should go with the quickfix for now and then think about how to stop the protocol from leaking into user code.

File: trino/client.py (hunk 2 above)
Comment (mdesmet) [reply trino/client.py:597]: 2 points here:

- It seems that there is a `TrinoResult` instantiation in  https://github.com/trinodb/trino-python-client/blob/aee6064c53536da6a73d0f2fa387eb2dfb222600/trino/client.py#L726. I'm not sure why that would be useful, as before fetching a query should be executed which will override the `_result` field in `TrinoQuery`.
- Trino's `QueryResults` can indeed expose a `null` data field, potentially on each fetch response, but that's handled in https://github.com/trinodb/trino-python-client/blob/aee6064c53536da6a73d0f2fa387eb2dfb222600/trino/client.py#L568

**Comment by mdesmet** [comment]
> @hashhar would you agree that to address the original issue we should add a fetch call after `scalar()` to drain the cursor? We might also want to document this somewhere. I don't think we should try to make the driver too smart to work around these protocol limitations.

This is exactly what sqlalchemy does when you `scalar_one` or `scalar_one_or_none`. Again we already fetch the next row in client module's `fetch` now, but even with fetching another record we have no guarantee that would finish the query (nextUri set to null) as @hashhar said.

https://github.com/sqlalchemy/sqlalchemy/blob/f8c4dba4e9f130c18ce00597c036bc26ae7abf90/lib/sqlalchemy/engine/result.py#L745-L748

File: trino/client.py [hunk 4]
```diff
@@ -697,6 +697,9 @@ def execute(self, additional_http_headers=None) -> TrinoResult:
         rows = self._row_mapper.map(status.rows) if self._row_mapper else status.rows
 
         self._result = TrinoResult(self, rows)
+        # Execute should block until at least one row is received
+        while not self.finished and not self.cancelled and len(self._result._rows) == 0:
+            self._result._rows += self.fetch()
```
Comment (hovaesco) [inline trino/client.py:702]: You could add `rows` getter property and setter to `TrinoQuery` to not directly access `_rows` and then change it also in `columns` property.

**Review by hovaesco — APPROVED** [review]
LGTM % https://github.com/trinodb/trino-python-client/pull/220#discussion_r970515709

**Review by hashhar — COMMENTED** [review]
This is a clever solution to the original problem. Thanks for coming up with the idea to make `execute` block for first response.

However this is also a breaking change (even though it improves experience) so I'm approving it but not merging until we release a version from current `master` so that people can upgrade to that version and then choose to decide whether they want to stay there for sometime before migrating to new blocking API.

And now that we are taking this direction it might be wise to decouple the Trino API handling from the db-api client and instead make available the fetched rows to the db-api cursor via a queue (list) instead of it directly fetching things from API. That gives future flexibility to introduce performance enhancements as well like the double-bufferring that the JDBC driver does for example and also make it possible to provide different cursor implementations.

**Comment by hashhar** [comment]
@findepi could you help with a release from current master (and also see my comment above)?

(I don't consider this issue a release blocker since it's a change in server behaviour only and as long as you call `fetchall` you're guaranteed to not run into this issue).

**Comment by findepi** [comment]
> could you help with a release from current master

sure, will coordinate over DM

**Comment by mdesmet** [comment]
> However this is also a breaking change (even though it improves experience) so I'm approving it but not merging until we release a version from current `master` so that people can upgrade to that version and then choose to decide whether they want to stay there for sometime before migrating to new blocking API.

Can you explain why you think this is a breaking change? IMHO the only way to break existing usage is if users didn't catch exceptions on` execute()`, which is a bug already as query submission can throw the same exceptions as `fetch*()`. The `fetch*()` operations continue to work as before, as proven by the integration tests.

Note that sqlalchemy doesn't call `fetch*()` on DML operations. Bringing this in would fix that.

> And now that we are taking this direction it might be wise to decouple the Trino API handling from the db-api client and instead make available the fetched rows to the db-api cursor via a queue (list) instead of it directly fetching things from API. That gives future flexibility to introduce performance enhancements as well like the double-bufferring that the JDBC driver does for example and also make it possible to provide different cursor implementations.

IMHO this is already decoupled. The `TrinoQuery` exists in client module and the exposes a lazy collection (an Iterator powered by a generator). I think this is the correct abstraction to use. I don't see leakage of the API details being introduced in this PR, actually the opposite is true: I would argue that in current code `execute()` is only query submission while `fetch*()` is query execution and result set scrolling, while this PR makes `execute()` query submission and execution and `fetch*()` result set scrolling which seems semantically more correct and in line with other dbapi implementations.

The [double buffering as in the java Trino client](https://github.com/mdesmet/trino/blob/8f27eca7ca34f172a95680d706c8c9427a5176b2/client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java#L235-L265), is implemented in this PR. Note that also the Java client doesn't use threading at this moment. I think it is a good idea to investigate but can be done independent from this PR.

I don't see why cursor implementations are impacted, cursors would take the `Iterable` and convert it for example in a dict instead of a tuple (many dbapi implementations have a `DictCursor`).

**Comment by hashhar** [comment]
> Can you explain why you think this is a breaking change? 

Because the API is now blocking.

>  I don't see leakage of the API details being introduced in this PR, actually the opposite is true:

I don't mean that this PR leaks the API details. I meant the opposite. Now we are one step closer to hide the API details within TrinoQuery and TrinoResult. An example of what this PR allows to do (but probably doesn't make sense) is to have a different impl of TrinoQuery which can probably use a different fictional transport mechanism to talk to Trino (instead of the REST API).

> The [double buffering as in the java Trino client](https://github.com/mdesmet/trino/blob/8f27eca7ca34f172a95680d706c8c9427a5176b2/client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java#L235-L265), is implemented in this PR.

True

> Note that also the Java client doesn't use threading at this moment. I think it is a good idea to investigate but can be done independent from this PR.

Exactly what I said above.

> I don't see why cursor implementations are impacted, cursors would take the Iterable and convert it for example in a dict instead of a tuple (many dbapi implementations have a DictCursor).

Again exactly what I said your PR allows us to do in future.

File: trino/sqlalchemy/dialect.py [hunk 5]
```diff
@@ -231,7 +231,7 @@ def get_view_definition(self, connection: Connection, view_name: str, schema: st
         """
         ).strip()
         res = connection.execute(sql.text(query), schema=schema, view=view_name)
-        return res.scalar()
+        return res.scalar_one_or_none()
```
Comment (hashhar) [inline trino/sqlalchemy/dialect.py:234]: can the sqlalchemy changes be a preparatory commit? Some context in the commit message will be useful (either as separate commit if possible or in current commit if not).

**Review by hashhar — APPROVED** [review]

**Comment by hashhar** [comment]
Newer Trino versions will include https://github.com/trinodb/trino/pull/14122 which can mean CI can be green without this change.

I think we should add one more entry to matrix with 395 as the version being tested for the meantime.

File: trino/sqlalchemy/dialect.py (hunk 5 above)
Comment (mdesmet) [reply trino/sqlalchemy/dialect.py:234]: Done

**Comment by mdesmet** [comment]
> Newer Trino versions will include [trinodb/trino#14122](https://github.com/trinodb/trino/pull/14122) which can mean CI can be green without this change.
> 
> I think we should add one more entry to matrix with 395 as the version being tested for the meantime.

I added the entry.

**Review by hashhar — APPROVED** [review]
[Acknowledge reception of data in TrinoResult](https://github.com/trinodb/trino-python-client/pull/220/commits/3f07f68402fab647d7cf3764407ae731e48589c8) fails without [Execute should block until at least one row is received](https://github.com/trinodb/trino-python-client/pull/220/commits/2539a2c1bcc3467907d1c55ef183c5395b1bb85f)

So it seems we need to squash those.

LGTM otherwise.

**Comment by mdesmet** [comment]
@hashhar: Squashed those commits. Are we good to go?

**Comment by hashhar** [comment]
Thanks. Good to go.

**Pull request merged**